clap_complete/src/dynamic/completer.rs

```diff
@@ -85,24 +90,33 @@ pub fn complete(
     ))
 }
 
+#[derive(Debug, PartialEq, Eq, Clone)]
+enum ParseState {
+    /// Parsing a value done, there is no state to record.
+    ValueDone,
+
+    /// Parsing a positional argument after `--`
+    Pos(usize),
+}
+
 fn complete_arg(
     arg: &clap_lex::ParsedArg<'_>,
     cmd: &clap::Command,
     current_dir: Option<&std::path::Path>,
     pos_index: usize,
-    is_escaped: bool,
+    state: ParseState,
 ) -> Result<Vec<(OsString, Option<StyledStr>)>, std::io::Error> {
     debug!(
-        "complete_arg: arg={:?}, cmd={:?}, current_dir={:?}, pos_index={}, is_escaped={}",
+        "complete_arg: arg={:?}, cmd={:?}, current_dir={:?}, pos_index={:?}, state={:?}",
         arg,
         cmd.get_name(),
         current_dir,
         pos_index,
-        is_escaped
+        state
     );
     let mut completions = Vec::new();
 
-    if !is_escaped {
+    if let ParseState::ValueDone = state {
```

Comment:
Seems like this should be a `match` and only for `ValueDone` state do we inference, otherwise we only respect the specified `ParserState`

ie
- `pos_index` should only be read within the `ValueDone` state
- `complete_subcommand` should only be called in the `ValueDone` state (atm, we might revisit this later)